decksite/data/rule.py

```diff
@@ -0,0 +1,189 @@
+    deck_ids: List[str] = []
+    rule_archetypes = {}
+    for r in (Container(row) for row in db().select(sql)):
+        deck_ids.append(str(r.deck_id))
+        rule_archetypes[r.deck_id] = r.rule_archetype_name
+    if len(deck_ids) == 0:
+        return []
+    result = deck.load_decks(where='d.id IN ('+','.join(deck_ids)+')')
+    for d in result:
+        d.rule_archetype_name = rule_archetypes[d.id]
+    return result
+
+def doubled_decks() -> List[Deck]:
+    sql = """SELECT
+                    deck_id, GROUP_CONCAT(archetype.name) AS concat_archetypes
+                FROM
+                    ({apply_rules_query}) AS applied_rules
+                JOIN
+                    archetype
+                ON
+                    applied_rules.archetype_id = archetype.id
+                GROUP BY
+                    deck_id
+                HAVING
+                    COUNT(DISTINCT archetype.id) > 1""".format(apply_rules_query=apply_rules_query())
+    deck_ids: List[str] = []
+    concat_archetypes = {}
+    for r in (Container(row) for row in db().select(sql)):
+        deck_ids.append(str(r.deck_id))
+        concat_archetypes[r.deck_id] = r.concat_archetypes
+    if len(deck_ids) == 0:
+        return []
+    result = deck.load_decks(where='d.id IN ('+','.join(deck_ids)+')')
+    for d in result:
+        d.concat_archetypes = concat_archetypes[d.id]
+    return result
+
+def load_all_rules() -> List[Container]:
+    result = []
+    result_by_id = {}
+    sql = 'SELECT rule.id as id, rule.archetype_id, archetype.name as archetype_name FROM rule JOIN archetype on rule.archetype_id = archetype.id'
+    for r in (Container(row) for row in db().select(sql)):
+        result.append(r)
+        result_by_id[r.id] = r
+        r.included_cards = []
+        r.excluded_cards = []
+    sql = 'SELECT rule_id, card, include FROM rule_card'
+    for r in (Container(row) for row in db().select(sql)):
+        if r.include:
+            result_by_id[r.rule_id].included_cards.append(r.card)
+        else:
+            result_by_id[r.rule_id].excluded_cards.append(r.card)
+    return result
+
+def add_rule(archetype_id: int) -> None:
+    sql = 'INSERT INTO rule (archetype_id) VALUES (%s)'
+    db().insert(sql, [archetype_id])
+
+def update_cards(rule_id: int, inc: str, exc: str) -> None:
+    db().begin('update_rule_cards')
+    sql = 'DELETE FROM rule_card WHERE rule_id = %s'
+    db().execute(sql, [rule_id])
+    for card in inc:
+        sql = 'INSERT INTO rule_card (rule_id, card, include) VALUES (%s, %s, TRUE)'
+        db().execute(sql, [rule_id, card])
+    for card in exc:
+        sql = 'INSERT INTO rule_card (rule_id, card, include) VALUES (%s, %s, FALSE)'
+        db().execute(sql, [rule_id, card])
+    db().commit('update_rule_cards')
+
+# Currently we do this query several times in a row, but at least with a small number of rules it's cheap enough not to matter
+def apply_rules_query(deck_query: str = '1 = 1'):
+    return f"""
```

Comment:
If you don't want leading whitespace on your SQL you can use the normal indentation for where you are in the code but then call `textwrap.dedent` on the resulting string.